pytorch_lightning/metrics/metric.py

```diff
@@ -37,21 +37,23 @@ class Metric(nn.Module, ABC):
     Args:
         compute_on_step:
             Forward only calls ``update()`` and returns None if this is set to False. default: True
-        ddp_sync_on_step:
+        dist_sync_on_step:
             Synchronize metric state across processes at each ``forward()``
-            before returning the value at the step. default: False
+            before returning the value at the step. Optionally accepts a callback that
+            performs the allgather operation on the metric state. When `True`, DDP
+            will be used to perform the allgather. default: False
         process_group:
             Specify the process group on which synchronization is called. default: None (which selects the entire world)
     """
     def __init__(
         self,
         compute_on_step: bool = True,
-        ddp_sync_on_step: bool = False,
+        dist_sync_on_step: Union[bool, Callable] = False,
```

Comment:
The logic here will need to be seperated. We will *always* want to sync when `.compute()` is called, this flag states whether we should sync on every single step.

`dist_sync_on_step` should still be a `bool`, but I think there needs to be an additional `dist_sync_fn` flag. This is why some of the tests are failing, as compute is giving incorrect values because it is not being synced.